src/main/scala/org/allenai/scholar/metrics/metadata/GrobidParser.scala

```diff
@@ -0,0 +1,196 @@
+    MetadataAndBibliography(metadata = extractMetadata(doc), bibs = extractBibEntriesWithId(doc).map(_._2))
+  }
+
+  private def extractMetadata(doc: Element) = PaperMetadata(
+    title =
+      Title(doc.findText("teiHeader>fileDesc>titleStmt>title")),
+    authors =
+      doc.select("teiHeader>fileDesc>sourceDesc>biblStruct>analytic>author").flatMap(author),
+    year =
+      doc.findAttributeValue("teiHeader>fileDesc>sourceDesc>biblStruct>monogr>imprint>date[type=published]", "when").extractYear,
+    venue =
+      Venue(doc.findText("teiHeader>fileDesc>sourceDesc>biblStruct>monogr>title").toTitle)
+  )
+
+  private def extractBibEntriesWithId(doc: Element) =
+    for {
+      bib <- doc.select("listBibl>biblStruct")
+    } yield {
+      val id = bib.attr("xml:id")
+      val title = bib.findText("analytic>title[type=main]").toTitle match {
+        case "" => Title(bib.findText("monogr>title"))
+        case s => Title(s)
+      }
+      val authors = bib.select("analytic>author").flatMap(author).toList match {
+        case List() => bib.select("monogr>author").flatMap(author).toList
+        case l => l
+      }
+      val venue = Venue(bib.findText("monogr>title"))
+      val yr = year(bib)
+      (id, PaperMetadata(title = title, authors = authors, venue = venue, year = yr))
+    }
+
+  private def extractSectionInfo(div: Element) = {
+    val bodyPlusHeaderText = div.text
+
+    val head = div.select("head").headOption
+    val (id, headerText, bodyTextOffset) = head match {
+      case Some(h) =>
+        val hText = h.text
+        (
+          h.attr("n").ifNonEmpty,
+          Some(hText),
+          hText.size + bodyPlusHeaderText.drop(hText.size).takeWhile(_ <= ' ').size
+        )
+      case None =>
+        (None, None, 0)
+    }
+    val section = Section(id = id, text = bodyPlusHeaderText.drop(bodyTextOffset), header = head.map(_.text))
+    (div, bodyPlusHeaderText, bodyTextOffset, section)
+  }
+
+  def extractStructuredDocument(xmlString: String): StructuredDocument = {
+    val doc = Jsoup.parse(xmlString, "", org.jsoup.parser.Parser.xmlParser())
+    val metadata = extractMetadata(doc)
+    val bibs = extractBibEntriesWithId(doc)
+    val paperAbstract = doc.select("teiHeader>profileDesc>abstract").headOption.map(_.text)
+    val sectionInfo = doc.select("text>body>div").map(extractSectionInfo)
+    val bibMentions =
+      for {
+        ref <- doc.select("ref[type=bibr")
+        ((div, fullText, offset, _), sectionNumber) <- sectionInfo.zipWithIndex.find {
+          case ((div, fullText, offset, _), i) => ref.parents.contains(div)
+        }
+      } yield {
+        val id = ref.attr("target").dropWhile(_ == '#')
+        val begin = ref.textOffset(div) - offset
+        val end = begin + ref.text.size
+        (id, Mention(sectionNumber, begin, end))
+      }
+    val mentionsOfBibEntry = bibMentions.toVector.groupBy(_._1).mapValues(_.map(_._2))
+    val bibsWithMentions = for ((id, bib) <- bibs) yield {
```

Comment:
`bibsWithMentions` is a `mutable.Buffer` which by default is a `mutable.ArrayBuffer` and thus has `IndexedSeq` as a trait. I don't think you need to call `toVector` here - it will needlessly cause copy/GC.